SaraAlert/SaraAlert
# SARAALERT-520: Adding SMART on FHIR Backend Processes Flow to API

# Description
Jira Ticket: [SARAALERT-520](https://tracker.codev.mitre.org/browse/SARAALERT-520)

This PR adds functionality to the existing API to support a workflow that does not require an end user to be involved. In other words, machine-to-machine API support. It follows the SMART on FHIR protocol outlined [here](https://hl7.org/fhir/uv/bulkdata/authorization/index.html). 

The API documentation has been updated with details - read the appropriate section there as well as the protocol linked above for context first.

**For this implementation, the approach was to update our existing Doorkeeper configuration to support a new workflow for machine-to-machine API access (the backend services protocol). This involved doing the following:**
-  Using Doorkeeper's custom handler functionality described here to carry out the necessary validation for the new workflow. 
- Adding three new columns to an existing Doorkeeper database Applicationtable to support client applications that want to use this workflow. 
- Updating existing API controller to scope access for client's using this workflow to the jurisdiction level, while fixing the access in the end-user workflow as well.

**IMPORTANT CHANGE TO EXISTING API WORKFLOW:**
- I noticed that no matter the user role, we were giving them access to all patients in their jurisdiction rather than all of the patients they have access to in the application. It has been updated such that if there is a current resource owner (end user), only allow access to patients that user has access to. The new workflow is at the jurisdiction level. 

**There are still some items to improve in future PRs:**
- Write tests for the Doorkeeper custom module.
- Support accepting and fetching from a client-hosted URL for the public key set for this workflow rather than require it be provided directly upon registration. This is the preferred approach as described in the link above.
- Update the UI to support creating applications for this workflow - as of now it has to be done through the console.
- BEFORE APPLICATIONS CAN BE REGISTERED ON PRODUCTION: When validating an incoming JWT assertion on a client request, verify that the associated jti (JWT ID) hasn't been seen in the last 5 minutes to prevent replay attacks. 
- Break out the API as a separate service in its own container.

There are tickets in the API epic that include all of the above.

**NOTE**: These changes are targeted for 1.12, but applications will not be registered on production (only demo) until after more testing, security review, and security-related updates (specifically the jti check mentioned above) have been included. We can inform jurisdictions that are interested in this workflow and tracking when it would be released that they can be registered on demo to pilot the new API workflow (as many have requested). 

# Important Changes
`app/controllers/fhir/r4/api_controller.rb`
- Added new `system` scopes.
- Updated to find accessable patients in two different ways to support both workflows. 1) If there is a current resource owner (end user), only allow access to patients that user has access to. This is a change - previously it was ALL patients in their jurisdiction. 2) If there is NOT a current resource owner (meaning no end user and following new workflow), get all patients in the client application's jurisdiction that was provided upon registration.
- Check for if the end user has API access (since that can currently been toggled) was not actually doing anything previously. This has been updated to not show any accessable patients if they do not have access toggled on.
- Uses the associated "shadow_user" for an application registered for the system flow when creating a patient. 

`config/credential_handler.rb`
- New module that handles the logic for the new workflow. This handler is automatically called when an accepted grant type request comes through. 

`config/initializers/doorkeeper.rb`
- Updated to use new module handler for extracting (and in some cases validating) client credentials.
- Updated to include new scopes.
- Updated to support new grant flow.

`db/migrate/20200905194603_update_doorkeeper_application_table.rb` and `db/schema.rb`
- Added three new columns to Doorkeeper OAuth Application table: `jurisdiction_id`, `public_key_set`, and `user_id` to support new workflow.

`test/controllers/fhir/r4/api_controller_test.rb`
- Updated existing tests to work with fix for api access per user.
- Added a bunch more tests.

`API.md`
- Significant updates to the API documentation.

`Gemfile`
- Added a new dependency to do safer YAML loading when deserializing the public key for this workflow. There can be risks when using this with ActiveRecord if it is used to replace the default YAML.load, but offers another approach for one time uses that is followed here. Read more [here](https://github.com/dtao/safe_yaml#what-if-i-dont-want-to-patch-yaml). Right now it will print a startup message until further configured (though we may not want to further configure it). 

# Testing
There is a lot of testing that can be done here, but fortunately no changes to the actual FHIR resource handling in the controller. To test the new workflow, I would recommend the following process to get started testing:
1. Create test application in the console using the same syntax as in the demo rake task. Make sure it has a jurisdiction_id, as well as a public_key_set that has been serialized. This public key set can be created like this: `public_key_set = {"keys": [public_key]}.deep_stringify_keys.to_yaml` where `public_key` is a JWK public key that can be generated [here](https://mkjwk.org/). 
2. Use postman to make an authorization request to the application (once it's running), using the `client_id` that you chose in step 1. This requires generated a JWT assertion following the steps described in the official protocol. In order to do so, use [this](https://jwt.io/) tool to generate the JWT assertion, making sure to use both the public and private key that were generated in step 1 (the PEM format) and adding the appropriate fields to the body and header. Follow the steps in the API.md documentation for this process. 
3. Verify that the authorization request went through and there was a 200 response with an access code - at this point you should be able to make normal API requests.
4. Make sure the appropriate status codes are returned if the request has issues that should be caught (expired signature, wrong iss, kid value that does not match the provided public key, etc.). 
5. Make sure the existing workflow also functions as it did before (except for the scope update as explained above).
